### PR TITLE
sdk: device_update: migrate to `azure_core::error::Error`

### PR DESCRIPTION
**Pull request by mfrw**
For:
- #771

Changes:

- Remove `thiserror` & replace with `azure_core::error::Error`

### CODE REVIEW

File: sdk/device_update/src/client.rs [hunk 1]
```diff
@@ -94,23 +98,28 @@ impl DeviceUpdateClient {
...
             return match headers.get("operation-location") {
-                Some(location) => location
-                    .to_str()
-                    .map(|x| x.to_string())
-                    .map_err(|_| Error::InvalidOperationPath),
-                None => Err(Error::NoOperationLocation),
+                Some(location) => location.to_str().map(|x| x.to_string()).map_err(|_| {
+                    Error::with_message(ErrorKind::Other, || {
+                        "successful import (202 status) but no operation-location header found"
+                            .to_string()
```
Comment (johnbatty) [inline sdk/device_update/src/client.rs:111]: This error needs to be swapped with the following one - i.e. this one should be "invalid characters..."

File: sdk/device_update/src/client.rs [hunk 2]
```diff
@@ -94,23 +98,28 @@ impl DeviceUpdateClient {
...
+                }),
+                None => Err(Error::with_message(ErrorKind::Other, || {
+                    "invalid characters in operation-location path".to_string()
+                })),
             };
         }
 
-        Err(Error::ImportError(resp.status()))
+        Err(Error::with_message(ErrorKind::Other, || {
+            format!("import unseccessful, status: {}", resp.status())
```
Comment (johnbatty) [inline sdk/device_update/src/client.rs:121]: typo: "unseccessful" -> "unsuccessful"

File: sdk/device_update/src/device_update.rs [hunk 3]
```diff
@@ -183,11 +186,27 @@ impl DeviceUpdateClient {
             let update_operation: UpdateOperation = self.get(uri.to_string()).await?;
 
             match update_operation.status {
-                OperationStatus::Failed => return Err(Error::ImportFailed(update_operation)),
+                OperationStatus::Failed => {
+                    return Err(Error::with_message(ErrorKind::Other, {
+                        || {
+                            format!(
+                                "import unsccessful with status failed. status: {:?}",
```
Comment (johnbatty) [inline sdk/device_update/src/device_update.rs:193]: typo: "unsccessful" => "unsuccessful"

**Review by johnbatty — COMMENTED** [review]
Looking good! Thanks very much for doing this.  Just a few minor markups.

File: sdk/device_update/src/client.rs (hunk 2 above)
Comment (mfrw) [reply sdk/device_update/src/client.rs]: Ahh :( my spellings :(

File: sdk/device_update/src/client.rs (hunk 1 above)
Comment (mfrw) [reply sdk/device_update/src/client.rs]: Sure

File: sdk/device_update/src/client.rs [hunk 4]
```diff
@@ -74,13 +75,16 @@ impl DeviceUpdateClient {
...
-            .map_err(|e| Error::Core(CoreError::Http(HttpError::ReadBytes(e.into()))))?;
-        serde_json::from_slice(&body).map_err(|e| Error::Core(CoreError::Json(e)))
+            .with_context(ErrorKind::Io, || {
+                format!("failed to send request. uri: {uri}")
+            })?;
+
+        let body = resp.bytes().await.with_context(ErrorKind::Io, || {
+            format!("failed to read response body text. uri: {uri}")
+        })?;
+        serde_json::from_slice(&body).with_context(ErrorKind::DataConversion, || {
```
Comment (cataggar) [inline sdk/device_update/src/client.rs:85]: You can use `.context` here.

File: sdk/device_update/src/client.rs [hunk 5]
```diff
@@ -74,13 +75,16 @@ impl DeviceUpdateClient {
...
+            "failed to deserialze json response body".to_string()
```
Comment (cataggar) [inline sdk/device_update/src/client.rs:86]: deserialize

File: sdk/device_update/src/client.rs [hunk 6]
```diff
@@ -94,23 +98,28 @@ impl DeviceUpdateClient {
...
 
         if resp.status() == 202u16 {
             let headers = resp.headers();
             return match headers.get("operation-location") {
-                Some(location) => location
-                    .to_str()
-                    .map(|x| x.to_string())
-                    .map_err(|_| Error::InvalidOperationPath),
-                None => Err(Error::NoOperationLocation),
+                Some(location) => location.to_str().map(|x| x.to_string()).map_err(|_| {
```
Comment (cataggar) [inline sdk/device_update/src/client.rs:108]: Use `context` instead of `map_err` to keep the source error.

File: sdk/device_update/src/client.rs [hunk 7]
```diff
@@ -94,23 +98,28 @@ impl DeviceUpdateClient {
...
+                    Error::with_message(ErrorKind::Other, || {
+                        "invalid characters in operation-location path".to_string()
+                    })
+                }),
+                None => Err(Error::with_message(ErrorKind::Other, || {
```
Comment (cataggar) [inline sdk/device_update/src/client.rs:113]: You are not formatting the error message, so you can just use `Error::message`

File: sdk/device_update/src/client.rs [hunk 8]
```diff
@@ -134,40 +144,57 @@ impl DeviceUpdateClient {
...
     use super::*;
+    use azure_core::error::{Result, ResultExt};
 
     #[test]
     fn can_extract_endpoint() -> Result<()> {
+        let url = "https://myadu.api.adu.microsoft.com";
         let suffix = extract_endpoint(
-            &Url::parse("https://myadu.api.adu.microsoft.com")
-                .map_err(|e| CoreError::Http(HttpError::Url(e)))?,
+            &Url::parse(url).with_context(ErrorKind::DataConversion, || {
```
Comment (cataggar) [inline sdk/device_update/src/client.rs:171]: These are tests, so context isn't really needed. You can just do `&Url::parse(url)?`. `From` is implemented.
https://github.com/Azure/azure-sdk-for-rust/blob/a6625a5f3b72cc6bf6095c7116d325a79e7816ce/sdk/core/src/error/mod.rs#L237-L241

Comment (cataggar) [reply sdk/device_update/src/client.rs:171]: May be check if `url::ParseError` provides the url that it fails on. If it does, then context probably isn't needed in a lot of these.

File: sdk/device_update/src/client.rs [hunk 9]
```diff
@@ -134,42 +144,48 @@ impl DeviceUpdateClient {
...
-        .ok_or(Error::DomainParse)?
-        .splitn(2, '.') // FIXME: replace with split_once() when it is in stable
-        .last()
-        .ok_or(Error::DomainParse)?;
+        .ok_or_else(|| {
+            Error::with_message(ErrorKind::DataConversion, || {
+                format!("could not get device update domain. url: {url}")
+            })
+        })?
+        .split_once('.')
```
Comment (mfrw) [inline sdk/device_update/src/client.rs:152]: I also noticed the `FIXME: use split_once`
Might as well do it as it is in stable now :)

File: sdk/device_update/src/client.rs (hunk 5 above)
Comment (mfrw) [reply sdk/device_update/src/client.rs]: 😢

File: sdk/device_update/src/client.rs (hunk 8 above)
Comment (mfrw) [reply sdk/device_update/src/client.rs]: It does not :(, but I think we should not clutter the tests.
I've added one test that should fail though.

**Review by bmc-msft — APPROVED** [review]

**Pull request merged**